tools/runfiles/runfiles.rs

```diff
@@ -53,23 +55,70 @@ enum Mode {
 #[derive(Debug)]
 pub struct Runfiles {
     mode: Mode,
+    repo_mapping: HashMap<String, String>,
+    source_repository: String,
 }
 
 impl Runfiles {
     /// Creates a manifest based Runfiles object when
     /// RUNFILES_MANIFEST_ONLY environment variable is present,
     /// or a directory based Runfiles object otherwise.
     pub fn create() -> io::Result<Self> {
-        if is_manifest_only() {
+        let bootstrap_runfiles = if is_manifest_only() {
             Self::create_manifest_based()
         } else {
             Self::create_directory_based()
-        }
+        };
+
+        bootstrap_runfiles.map(|r| {
+            let repo_mapping_file = r.rlocation("_repo_mapping");
+            if repo_mapping_file.exists() {
+                let repo_mapping = std::fs::read_to_string(repo_mapping_file)
+                    .expect("failed to read _repo_mapping file")
+                    .lines()
+                    .map(|line| {
+                        let triplet: Vec<&str> = line.splitn(3, ',').collect();
+                        (
+                            format!("{},{}", triplet[0], triplet[1]),
+                            triplet[2].to_string(),
+                        )
+                    })
+                    .collect::<HashMap<_, _>>();
+                Runfiles {
+                    mode: r.mode,
+                    source_repository: r.source_repository,
+                    repo_mapping,
+                }
+            } else {
+                Runfiles {
+                    mode: r.mode,
+                    source_repository: r.source_repository,
+                    repo_mapping: HashMap::new(),
+                }
+            }
+        })
+    }
+
+    fn get_source_repository() -> String {
+        let caller = std::env::current_exe().expect("failed to get current executable path");
```

Comment:
This looks problematic: What if a `rust_binary` in repo `foo` depends on a `rust_library` in repo `bar`? Lookup should be relative to `bar` in that case, but this would result in `foo`.

Maybe this is a use case for `std::env!` together with rules support that adds the repository name of the current target to the compilation environment when the target has a direct dependency on the runfiles library? This is very similar to what I did for `cc_*` rules.

Reply:
I think the correct approach would be to use a macro instead of a function. We need the compilation context of the caller.

Rlocation!(r, "foo") would expand to r.rlocation(foo, env!("workspace"))

Note that that environment variable is evaluated at compile time, not runtime. Not sure if it's the same idea that @fmeum had

Sorry for formatting, on phone

Reply:
@matts1 Yes, that's what I had in mind, a macro that itself uses `std::env!` to evaluate to the repository name of the target containing the `rlocation` call. Compile-time evaluation seems necessary as runtime evaluation wouldn't be able to distinguish the calls based on source file.

Reply:
Thanks for the input. I'll take a look if I can change the library so that it uses a macro. Are we fine with a breaking change? 

Just to sum up, the approach you both are suggesting is that I add the canonical workspace name as an env to the rustc compilation action and then use a macro that uses `std:env!` to embed the workspace name into the compiled binary that is using the macro.

That leaves the question if there is a good way to get the canonical workspace name within a rule? AFAIK `ctx.workspace_name` always  returns `_main` when bzlmod is enabled. It's possible to use `native.repo_name` but then all rules would have to be wrapped in a macro.

Reply:
The term "workspace name" has always been confusing and isn't really relevant with Bzlmod anymore as you noticed. Instead, you want the canonical repository name of the current target, which you can obtain via `ctx.label.repo_name` (available as of 6.5.0) or `ctx.label.workspace_name` (a terribly misnamed alias of `repo_name`, but available in all versions of Bazel).

Reply:
Alrighty, I'll use `ctx.label.workspace_name` for increased compatibility.